### ISSUE TITLE
JavaScript: understand "this"

### ISSUE THREAD

**Issue by sydneytiger** [sydneytiger/blog#4]
`this` binding is confusing, as the iconic feature is JavaScript. A experienced developer could fail in figure it out the true value of `this` in code.

## Call-site

**Call-site is where a function get called not where it's declared** Understanding call-site is the key of understanding `this` binding. `this` is referencing the call-site.

```javascript
function A(){
  B(); // function B get called, the call-site is A
}

function B(){
  C(); // function C get called, the call-site is B
}

function C(){
  ...
}

A(); // function A get called, the call-site is global
```

Refer to the demo code above, function A, B, C are declared in global scope but the `call-site` is different. It is determined by the called location.

## Four Rules

Now we understand `call-site`, then we have to know 4 binding rules.

1. Default Binding
2. Implicit Binding
3. Explicit Binding
4. New Binding

In practice, we must inspect the `call-site` of the code and determine which rule to apply, and then we can work out the `this` value.

## Default Binding

This is a binding rule that none of other rules can apply. It is the most common case: **Standalone function invocation**

```javascript
function A() {
    console.log(this.a);
    console.log(this.b);
}

var a = 1;

A();
// 1
// undefined
```

We notice that function `A()` is called, the `this.a` resolves to the global variable `a`. The `this.b` response with `undefined` as no `b` is declared in global scope. Now when we declare `b` and run `A()` again. What would happen.

```javascript
var b = 2;

A();
// 1
// 2
```

Therefore, in **default binding `this` is global**.

## Implicit Binding

The second rule is that the `call-site` is a context object like `obj.A`.
Consider:

```javascript
function A() {
    console.log(this.a);
    console.log(this.b);
}

var obj = {
    a: 1,
    b: 2,
    fn: A
};

var a = 'global';

obj.fn();
// 1
// 2
```

Well, function `A()` is declared in global and then added to `obj` as a property `fn`. `A()` is not contain or child to `obj`. It is just referenced.

However, when `obj.fn()` invokes, the `call-site` of `A()` is actually `obj`, not global. Implicit binding rule says `obj` should be used for `this` binding. So `obj.a` has been logged, not `global.a`;

Ok, what is the output of the code below:

```javascript
function A() {
    console.log(this.a);
    console.log(this.b);
}

var obj2 = {
    a: 1,
    b: 2,
    fn: A
};

var obj1 = {
    a: 3,
    b: 4,
    obj2: obj2
};

obj1.obj2.fn();
```

`obj1` references `obj2` and `obj2` reference `A()`. Should it outputting `1, 2` or `3, 4`. The answer is:

```javascript
// 1
// 2
```

In ab object property reference chain, the `call-site` is the last level object. Or in another word. **The object close to the function wins**

![image](https://user-images.githubusercontent.com/1787825/61573750-edcfb500-aaf6-11e9-9cac-6203c1686521.png)

Let's consider another example here:

```javascript
function A() {
    console.log(this.a);
    console.log(this.b);
}

var obj = {
    a: 1,
    b: 2,
    fn: A
};

var a = 'global a';
var b = 'global b';

var funcB = obj.fn;

funcB();
// global a
// global b
```

`funcB` references `obj.fn`, `obj.fn` reference `A()`, when `funcB` invokes, `obj` has nothing to do here, bot `a` and `b` are point to global variables. According to **The object close to the invoke function wins**, no object close to `funcB()` so it falls back to the **Default Binding**. `this` binds to global. It is also known as **Implicitly lost**.

![image](https://user-images.githubusercontent.com/1787825/61423146-0bafe500-a952-11e9-955a-eb71811b5074.png)

When parameter passing an implicit reference function, the **implicitly lost** occurs. Here is an examples.

```javascript
function A() {
    console.log(this.a);
}

function invoker(fn) {
    fn();
}

var obj1 = {
    a: 1,
    fn: A
};

var obj2 = {
    a: 2,
    invoker: invoker
};

var a = 6;

obj1.fn(); // 1 *implicit reference function

invoker(obj1.fn); // 6

obj2.invoker(obj1.fn); // 6

setTimeout(obj1.fn, 10); // 6
//pass implicit reference function as parameter or callback results in implicit lost. Use default binding
```

It is JavaScript nature that callback function loses its `this` binding. And we, unfortunately, are not able to change it. With ES6 arrow function, it save us from the **implicitly lost**

```javascript
function A() {
    console.log(this.a);
}

var obj1 = {
    a: 1,
    fn: A
};

setTimeout(() => {
    obj1.fn();
}, 10); // 1
// Hallelujah
```

## Explicit Binding

As we saw, `implicit binding` is confusing and we have to amend the context object to reference the invoke function. And when the function is called, we have to carefully investigate the call-site object.

To avoid that, we have `explicit binding`. It explicitly tells a function to run under a given context. They are:

1. `call()`
2. `apply()`
3. `bind()`

They are built in functions in the `Prototype` of Function. We do not need to mutate the context object. They all take the first argument as context object. The `this` inside the invoke function has explicitly bind to this context object and will not change.

#### Call & Apply

Let's consider

```javascript
function A() {
    console.log(this.a);
}

const obj = {
    a: 1
};

var a = 'global';

A.call(obj); // 1
```

We invoke `A()` with `explicit binding` by `A.call(obj)` which forces the `this` to be `obj` not global.

If passing a primitive value as first argument. The primitive value will be `boxing` to an object. e.g. string => new String(), boolean => new Boolean() and number => new Number().

```javascript
function A() {
    console.log(this);
}

const obj = {
    a: 1
};

A.call(obj); // {a:1}
A.call(2); // Number {2}
```

The `call()` accepting unlimited parameters, first one is context, from the second one onward, they will be passed to invoke function as arguments.

```javascript
function A(b, c) {
    console.log(this.a + b + c);
    console.log(b);
    console.log(c);
    console.log(arguments[2]);
    console.log(arguments[3]);
}

const obj = {
    a: 1
};

A.call(obj, 2, 3, 4);
// 6
// 2
// 3
// 4
// undefined
```

`apply()` is not difference to `call()` but accepting two parameters. the first one is the context object, the second one is the an array passing to invoke function as arguments.

```javascript
A.call(obj, 2, 3, 4);
A.apply(obj, [2, 3, 4]);
```

These two line of code outputting the same result.

#### Bind

`bind()` function takes one parameter only which is the context object and return a new function. Different from `call()` and `apply()`. It does not invoke the function. We have to invoke it manually later on.

```javascript
function A() {
    console.log(this.a);
}

const obj = {
    a: 1
};

const B = A.bind(obj);

B(); // 1
```

It is also known as `hard binding`. Once a function is bind to an object, wherever we invoke it, the `this` binding is always the `obj`. How about we call the `B()` with `call()`? Let's see the code:

```javascript
const obj2 = {
    a: 2
};

B.call(obj2); // 1
```

We `call` the `B()` by giving a context `obj2`. Since `B()` has been `bind()` to `obj` previously. The `obj2` will not take into effect.

## `new` Binding

In object oriented programming language, the `new` keyword calls the `constructor` function of the class and returns an instance of the class. JavaScript has `new` keyword too, however, totally different from the class-oriented language.

The `constructor` function in JavaScript is just a regular function. Calling it will not return any instance of class. It is not even a special type of function. So what happen under the hood of `new` a function?

1. create an empty context object.
2. link the context object **proto** to constructor prototype.
3. execute the constructor function with the context object.
4. unless the constructor reruns an object or a function, otherwise;, return the execute result.

Consider:

```javascript
function A(a, b) {
    this.a = a;
    this.b = b;
}

const B = new A(6);
console.log(B.a); // 6
console.log(B.b); // undefined

const C = new A(1, 2);
console.log(C.a); // 1
console.log(C.b); // 2
```

By calling `A()` with `new` keyword. It created a new context object and set the `this.a` and `this.b` by executing `A()`. Now the `this` bound to the context object. This is called `new binding`.

## Priority

We have walked through the 4 rules for binding of `this` in function call. What happen if multiple rule may apply to? There is which rule get the priority?

First of all **`default binding` take the lowest priority**.

Let's find out the precedence of `implicit binding`, `explicit binding` and `new binding`.

Try this one:

```javascript
function A() {
    console.log(this.a);
}

const obj1 = { a: 1, fn: A };
const obj2 = { a: 2, fn: A };

// implicit binding
obj1.fn(); // 1
obj2.fn(); // 2

// explicit binding
A.call(obj2); // 2
A.call(obj1); // 1

// combine
obj1.fn.call(obj2); // 2
obj2.fn.call(obj1); // 1
```

`obj1.fn.call(obj2)` combines the `implicit binding` and `explicit binding`. The results are the same as 'explicit binding'. It concludes that **`explicit binding` takes precedence over `implicit binding`**

How about `new binding`? Consider the code:

```javascript
function A(a) {
    this.a = a;
}

const obj1 = { fn: A };

const obj2 = {};

// implicit binding
obj1.fn('implicit');
console.log(obj1.a); // implicit

// explicit binding
A.call(obj2, 'explicit');
console.log(obj2.a); // explicit

// new binding vs implicit binding
const B = new obj1.fn('new');
console.log(obj1.a); // implicit
console.log(B.a); // new

const C = new A.call(obj2, 'new'); // TypeError: A.call is not a constructor
```

The result tells us `new binding` has higher priority than `implicit binding`. Unfortunately, we cannot do `new A.call(...)`. Please checkout another sample:

```javascript
const obj3 = {};

const D = A.bind(obj3);

// explicit biding
D('explicit');
console.log(obj3.a); // explicit

// new binding vs explicit binding
const E = new D('new');
console.log(obj3.a); // explicit
console.log(E.a); // new
```

`D()` as `hard bind` to obj3, but `new D()` does not change the `obj3.a`. The `new` create brand new object `E` with value of 'new' without effecting `obj3.a`.

In summary, the priority order of the 4 rules is:
**`new binding` > `explicit binding` > `implicit binding` > `default binding`**
